_posts/2021-05-28-why-avoid-an-orm.md

```diff
@@ -0,0 +1,231 @@
+ 
+If you update the department of an employee by saying `employee.setDepartment(newDepartment)`, the modification is probably going to happen as was intended, but what if you say:
+`department.addEmployee(newEmployee);`? Is the change going to be persisted?
+
+Probably not, since the ORM would need to implement behind the scenes some List-like collection which understood that additions and removals need to be handled with separate updates to another database table. Who knows if your ORM does this?
+
+What if I'm already doing something with an instance of employees collection of a department, and I then set the department of one of the employees to something else:
+```java
+for (Employee emp: department.getEmployees())
+    emp.setDepartment(anotherDepartment);
+}
+```
+
+For the object model to remain internally consistent the employee would suddenly need to disappear from the collection, and now we are in the rabbit hole of concurrency and concurrent modifications to a collection possibly under iteration. Let's not go there.
+ 
+Since the ORMs don't do this (at least I think none do), you could implement `setDepartment` and `addEmployee` methods yourself correctly so that they always work. You'd need the `getEmployees()` method to always return either an immutable or some [copy-on-write](https://en.wikipedia.org/wiki/Copy-on-write) collection, so it's not going to be 100% in sync. Also, if you happen to make copies, they won't be updated, but hey, it's at least almost perfect, right? Maybe, but no one's going to implement that since it's just too complex. Easier to just forbid add/remove methods and handle the relationship from the single side (`setDepartment`). And thus we've lost another part of our clean object model.
+ 
+There's still the problem of keeping the sides in sync, though. And the results aren't even (locally) deterministic. If you set the department of an employee via `employee.setDepartment(newDep)`, and after that call `department.getEmployees()`, the new employee is going to be found in the collection
+- unless the employees were configured to be eagerly fetched
+  - in which case the collection was already populated when the department was fetched, and won't be re-populated automatically
+- unless the previous code in the previous lines (or the previous functions) had already called `department.getEmployees()` at least once
+  - in which case the collection was already populated before the modification, and will be returned in the original state
+- unless some earlier code within the same session (~transaction) had already called `getEmployess()` -method of the same department which might have been a different department object retrieved from `em.get(depId)` but which at least Hibernate caches in memory and returns the same object instance on subsequent calls within the same transaction
```

Comment:
This sentence would benefit from splitting.

Reply:
Yes, this is an awkward list, but I don't really have ideas how to improve it, since I think I don't think I want to turn it into just regular text...